chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -818,53 +818,89 @@ impl PeerManagerActor {
                     NetworkResponses::RouteNotFound
                 }
             }
-            NetworkRequests::StateRequestHeader { shard_id, sync_hash, peer_id } => {
-                if self.state.tier2.send_message(
-                    peer_id,
-                    Arc::new(PeerMessage::StateRequestHeader(shard_id, sync_hash)),
+            NetworkRequests::StateRequestHeader { shard_id, sync_hash, sync_prev_prev_hash } => {
+                // The node needs to include its own public address in the request
+                // so that the response can be sent over a direct Tier3 connection.
+                let Some(addr) = *self.state.my_public_addr.read() else {
+                    return NetworkResponses::AwaitingIpSelfDiscovery;
+                };
+
+                // Select a peer which has advertised availability of the desired
+                // state snapshot.
+                let Some(peer_id) = self
+                    .state
+                    .snapshot_hosts
+                    .select_host_for_header(&sync_prev_prev_hash, shard_id)
+                else {
+                    tracing::debug!(target: "network", ?shard_id, ?sync_hash, "no snapshot hosts available");
+                    return NetworkResponses::NoDestinationsAvailable;
```

Comment:
Will it be retried later?

Reply:
Yes, the state sync actor handles the responsibility of trying again later. From the network crate's point of view, a NetworkRequests triggers a single best effort attempt.

Reply:
Yes, the state sync actor handles the responsibility of trying again later. From the network crate's point of view, a NetworkRequest triggers a single best effort attempt.